README.md
# Mixpanel Data Export 
This tool has been written in Ruby and packaged as a project.  It is not a single script but rather, a set of utilities that work together to provide data export functionality. It is a means for exporting all of your event data from Mixpanel using their published API.

Note that while this tool was designed to help you export your event data from Mixpanel, it was not built by Mixpanel. 

_All instructions that follow assume that you are in the project's root directory._

## Installation

You should have Ruby 2.x installed, as well as Rubygems, including the Bundler gem. For help installing Ruby on your system, see [the official Ruby installation page](https://www.ruby-lang.org/en/documentation/installation/). For help with installing RubyGems, see [RubyGems Basics](http://guides.rubygems.org/rubygems-basics/).

You'll then use [Bundler](http://bundler.io) to install the gems used by this project by issuing the command:

`bundle install`

Next, you'll copy the config files bundled as templates, removing the `.template` token in your copy. These files are located in the project's `config` folder.

* Exporter.config.*template*.json → Exporter.config.json
* SSL.config.*template*.json → SSL.config.json
* MixpanelCredentials.config.*template*.json → MixpanelCredentials.config.json

## Configuration

Before you can run the script, you need to configure the three JSON files you just made copies of.

#### File: Exporter.config.json

In this file, you specify in which directory the download-generated CSV files should be placed, using the entry `Output Directory`.

You also specify the range of days in which data should be fetched from Mixpanel.

```
{
  "Output Directory": "~/Desktop/App Output/Mixpanel Export",
  "From Days Ago": 1,
  "To Days Ago": 0,
  "Create Sub-folders by Date": false,
  "Use Event Dates in Filenames" : true,
  "Events to Exclude" : ["Data Imported", "Notification Received"],
  "Events to Include" : [],
  "Scale Request Expiry" : true,
  "Scale Request Seconds Per Day" : 1.3,
  "Unscaled Request Expiry Duration in Seconds" : 80,
  "Maximum Scaled Request Expiry Duration in Seconds" : 200
}
```

The `From Days Ago` is for how many days back the **start** of the export range is.

The `To Days Ago` is for how many days back the **end** of the export range is.

The `Create Sub-folders by Date` takes a boolean value. We recommend setting this to `false`, so that all of the event data across multiple dates, sits inside a single folder. If you do set this to `true`, then files we be placed in subfolders named after the date of download.

The `Use Event Dates in Filenames` determines whether we use the date range of the events being downloaded, in the actual filenames of the downloaded event data. We recommend setting this to `true`. Doing so will give you filenames with both the start and end date in the filename. Setting this to `false` means that the only date in the filename will be the date that the file was downloaded.

The `Events to Exclude` is an optional array of the names of events that you do *not* want to be downloaded. If you provide a non-empty array here, we'll also ignore the value you set for the mutually exclusive property `Events to Include`.

The `Events to Include` is mutually exclusive with the property `Events to Exclude`. If you provide a non-empty array for this property, we'll *only* download those events.

The `Scale Request Expiry` has a Boolean setting. We recommend setting this to `true`. Doing so, allows us to scale the request expiry timeout to be proportionate to the number of days you are downloading data for. Asking for data spanning weeks or months will take longer than asking for data from just yesterday. We don't unilaterally set a high number for the request expiry, because doing so is a potential security risk; anyone else who sniffs the request URL could re-issue it to obtain your data. Mixpanel [tutorials](https://mixpanel.com/docs/api-documentation/a-tutorial-on-exporting-data) generally use 10 minutes (600 seconds). We err on the side of being more conservative.

The `Scale Request Seconds Per Day` lets you define how generous a timeout you desire, in a formulaic manner. For example, if you set this to 0.5 and issue a query spanning 200 days, you will then have set an expiry of 100 seconds. Note that the tool has a minimum request expiry duration of 30 seconds, since most requests, event for just a single day's worth of data, take several seconds.

The `Unscaled Request Expiry Duration in Seconds` is what the tool will consult if you set  `Scale Request Expiry` to `false`. It is an absolute number that will be used across the board for request expiry. If you go this route, we suggest using `60` seconds if you plan to make requests spanning less than a few days. For requests that can span months, you should use something closer to `400` seconds. The Mixpanel data export web service can be really slow when querying data older than a few days.

The `Maximum Scaled Request Expiry Duration in Seconds` is the cap we should apply to the request expiry duration when `Scale Request Expiry` is set to `true`. This protects the tool from inadvertently calculating an expiry duration into hours (creating a security risk). Instead, we'll always cap the request expiry duration at this value.

The recommended configuration is to use the defaults shown above for `From Days Ago` and `To Days Ago`, where you are always retrieving *yesterday's* data. 

Schedule this script to run daily, and you'll always download the latest data available, without any gaps in your collected data.

Should you need to, you can of course edit the date range in this file and export event data for a different span of time.

#### File: SSL.config.json

In order for this Mixpanel Data Export Tool tool to talk to Mixpanel's servers over SSL, we have to advise the tool on where it can find the bundle of root certificates from various Certificate Authorities.

On many systems, these are found in a file named `cacert.pem`. In the example below, the path used is for the OpenSSL bundled root certificates file.

This may be different on Windows and Linux systems.

```
{
  "certificate_authority": {
    "cert_file_path": "/usr/local/etc/openssl/certs/cacert.pem"
  }
}
```

For more information on root certificates, you may find this link helpful: [http://wiki.cacert.org/FAQ/ImportRootCert](http://wiki.cacert.org/FAQ/ImportRootCert).


#### File: MixpanelCredentials.config.json

This file is where you store your API credentials to access your data on Mixpanel.

The contents shown below are fake keys that don't actually work -- you need to provide your own.

```
{
  "API Key": "92a533a1284ee3012e31d2a3e7bff8da",
  "API Secret": "0a12345678f165a2312a72b9e4331288"
}
```

Per the Mixpanel documentation on their Data Export API:

> Note: your **api_key** and **api_secret** can be found in your [account](https://mixpanel.com/account/) menu within the Projects tab.


## Usage
It is recommended that you schedule this tool to run nightly. That way, you are at most, just one day away from having all the data that Mixpanel has collected.

`ruby ./lib/mixpanel_data_export.rb`

An illustration of what to expect when you run this command is below. Of course, your events will differ in name, number and quantity of data retrieved.

![usage example](https://github.com/appstronomy/mixpanel-data-export/blob/master/Mixpanel%20Data%20Export.gif)

## Output Files

If you set `Create Sub-folders by Date` to `true`, then a new directory is created for each day that the script is run. For example, if you run the script on July 30, 2016, you'll see a directory created with the name '2016-07-30'.

By default however, all files are dropped into your output directory, without any sub-folder segmentation.

A sub-directory of your output directory is also created if need be, called 'logs'. Each day's run appends to a dated log file. 

Each downloaded file contains the name of the event it holds data for.

If you set `Use Event Dates in Filenames` to `true`, then your downloaded event data will be named  with the date range reflecting the date range of the event data itself. If you set this to `false`, then the date of the download is what you'll find in the filename.

Each event creates its own CSV file with data for just that event. If no data was downloaded for a particular event, then no CSV file is created for that event.

The CSV file has columns sorted alphabetically, with the exception of the first column, which will always include the name of the event (the column is called *event*). The rest of the columns follow Ruby's standard sorting algorithm; special characters and numbers come first, then capital letters and then lower case letters.

## Development
### Programmer Documentation
You can find module and class documentation in the `/doc` folder. Point your browser to the `index.html` file to get started.

To regenerate the documentation, make sure you have the `yard` gem installed, and then from the root of this project, issue the `yardoc` command.

### Enhancements
Please open a new issue to discuss changes you'd like to make. We can coordinate before you make a pull request.

### To-do
This project does need unit tests. We only have placeholder files for [rspec](http://rspec.info/) based testing.

## Troubleshooting

If you have any issues installing the gems via Bundler, you can try using Ruby 2.1.2, which was used to successfully compile finicky gems like 'redcarpet' on a Mac.

## Changelog

See the [Changelog](https://github.com/appstronomy/mixpanel-data-export/blob/master/Changelog.md) for a history of release notes.

## Collaborators and Maintainers

See the [Contributors](https://github.com/appstronomy/mixpanel-data-export/graphs/contributors) page for details.

## Copyright 

Copyright (c) 2015-2016+ [Appstronomy](http://appstronomy.com/), LLC. See [license](https://github.com/appstronomy/mixpanel-data-export/blob/master/LICENSE.txt) for details.


Changelog.md
## 0.6.0 (2016-02-26)

This version of the tool enhances the configuration subsystem, adding configuration keys:

  * Create Sub-folders by Date
  * Use Event Dates in Filenames
  * Events to Exclude
  * Events to Include
  * Scale Request Expiry
  * Scale Request Seconds Per Day
  * Unscaled Request Expiry Duration in Seconds
  * Maximum Scaled Request Expiry Duration in Seconds


## 0.5.0 (2015-07-30)

This is the initial version of the tool.

Workarounds:

  - Not Applicable

Bug Fixes:

  - None


spec/spec_helper.rb
require_relative '../lib/mixpanel_data_export'

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/util_connection_spec.rb
require_relative 'spec_helper'


describe 'Connection' do

  before :each do
    @connection = Util::Connection.new('http://mixpanel.com')
  end


  it 'should be an instance of Util::Connection' do
    expect(@connection).to be_an_instance_of Util::Connection
  end

end

spec/mixpanel_service_spec.rb
require 'rspec'

describe 'The Mixpanel Service' do

  before :each do
    @service = Mixpanel::Service.new
  end

  it 'should should be an instance of Mixpanel::Service' do
    expect(@service).to be_an_instance_of Mixpanel::Service
  end

end

lib/util/connection.rb
# Mixpanel Ruby Data Export Tool
#
# Copyright (c) 2015+ Appstronomy, LLC.
# See LICENSE.txt for details on use.

# This module contains supporting utilities for the classes in the Mixpanel module,
# including helper files to assist with loadpath setup and logging.
module Util

  require 'rubygems'
  require 'bundler/setup'
  require 'net/http'
  require 'net/https'
  require 'uri'
  require 'ostruct'
  require 'json'
  require 'logging'

  # The Connection class handles the underlying HTTP and HTTPS communication to whichever
  # endpoint and path is requested.
  #
  # Credit
  # Several ideas were derived from http://danknox.github.io/2013/01/27/using-rubys-native-nethttp-library/
  #
  # @author Sohail Ahmed, https://github.com/idStar
  class Connection

    # Constants

    DEFAULT_CA_CERT_PATH = '/usr/local/etc/openssl/certs/cacert.pem'

    VERB_MAP = {
        :get    => Net::HTTP::Get,
        :post   => Net::HTTP::Post,
        :put    => Net::HTTP::Put,
        :delete => Net::HTTP::Delete
    }

    # Instance Variables

    # The endpoint that we represent a connection to.
    # For example, 'https://data.mixpanel.com' or 'http://myserver.com'.
    # @return [String]
    attr_reader :endpoint

    # Our connection object. May actually be HTTPS if the endpoint
    # that we were initialized with used 'https://'.
    # @return [Net::HTTP]
    attr_reader :http

    # The full path (including the actual .pem file) that we use for the
    #   root Certificate Authority with which SSL connections are verified.
    # @return [String]
    attr_reader :ca_cert_path


    # Methods

    # Sets up our Connection with a pre-configured @http object that knows the endpoint
    # and protocol to use -- whether http or https.
    #
    # @param  [String] ca_cert_path Optional. The path to the 'cacert.pem' file.
    #         Not needed if the default is fine, or you are not sending in an SSL endpoint, or if
    #         you have provided a value in the SSL.config.json file to use instead.
    def initialize(endpoint, ca_cert_path = DEFAULT_CA_CERT_PATH)
      configure_logging
      @endpoint = endpoint
      uri = URI.parse(@endpoint) # This will correctly determine port based on protocol scheme
      @http = Net::HTTP.new(uri.host, uri.port)

      # Where we given a secure endpoint?
      if uri.scheme == 'https'
        @ca_cert_path = ca_cert_path
        load_ssl_config if ca_cert_path == DEFAULT_CA_CERT_PATH # If not specified by the caller, load the config file
        configure_with_ssl @ca_cert_path
        @logger.info "Using SSL for endpoint '#{@endpoint}' with CA certificates path: #{@ca_cert_path}."
      end

      @logger.info 'Initialized Connection instance.'
    end


    def get(path, params, return_json = false)
      route_request :get, path, params, return_json
    end


    def post(path, params, return_json = false)
      route_request :post, path, params, return_json
    end


    def put(path, params, return_json = false)
      route_request :put, path, params, return_json
    end


    def delete(path, params, return_json = false)
      route_request :delete, path, params, return_json
    end


    private

    # Sets up our logging instance attribute, with a logger for this class.
    # By default, this class specific logger inherits settings from the root logger.
    # @return [void]
    def configure_logging
      @logger = Logging.logger[self]
    end


    # We'll attempt to load an `SSL.config.json` file, and if we find it, parse it and
    # retrieve a file path for Certificate Authority certs, we'll set that as
    # {#ca_cert_path}.
    def load_ssl_config
      file_path = File.join(__dir__, '../../config/SSL.config.json')

      if File.exists? file_path
        file = File.read(file_path)
        config_info = JSON.parse(file)

        retrieved_path = config_info['certificate_authority']['cert_file_path']
        @ca_cert_path = retrieved_path if retrieved_path
      end
    end


    # Configures our @http object to use SSL with a verified peer, pointing the x509
    # certificate store directly at the provided Certificate Authority path.
    #
    # @param ca_cert_path The explicit full path to the 'cacert.pem' file that contains the root CA
    def configure_with_ssl(ca_cert_path)
      # Ensure the http object uses SSL and does so with a verified peer:
      @http.use_ssl = true
      @http.verify_mode = OpenSSL::SSL::VERIFY_PEER

      # Point the http object to the certificate store explicitly, so
      # we're not at the mercy of it not being found on the host system:
      @http.cert_store = OpenSSL::X509::Store.new
      @http.cert_store.set_default_paths
      @http.cert_store.add_file(ca_cert_path)
    end


    def route_request(method, path, params, return_json)
      request_method = return_json ? 'request_json' : 'request'
      send(request_method, method, path, params)
    end


    def request_json(method, path, params)
      response = request(method, path, params)
      body = JSON.parse(response.body)

      OpenStruct.new(:code => response.code, :body => body)
    rescue JSON::ParserError
      @logger.error "Could not parse response into JSON. Original request info: \
        {method = #{method}, path = #{path}, params = #{params}}"
      response
    end


    def request(method, path, params = {})
      case method
        when :get
          full_path = encode_path_params(path, params)
          @logger.debug "request full path is: #{full_path}"
          request = VERB_MAP[method.to_sym].new(full_path)
        else
          request = VERB_MAP[method.to_sym].new(path)
          request.set_form_data(params)
      end

      @http.request(request)
    end


    def encode_path_params(path, params)
      if params.length > 0
        # Encode parameters suitable for a web query string:
        encoded = URI.encode_www_form(params)
        # Separate the path from the encoded params with a query string operator:
        [path, encoded].join('?')
      else
        path
      end
    end



  end

end


# ---------- Development Testing ----------

# connection = Appstronomy::Connection.new('http://www.yahoo.ca')
# response = connection.get '.', {}
# puts response.body

lib/mixpanel/service.rb
# Mixpanel Ruby Data Export Tool
#
# Copyright (c) 2015+ Appstronomy, LLC.
# See LICENSE.txt for details on use.

module Mixpanel

  require 'rubygems'
  require 'bundler/setup'
  require 'active_support/all' # For mixin conveniences handling time and numbers
  require_relative '../util/connection'

  # Handles creating a secure connection to the Mixpanel API Service and
  # providing responses to predefined Mixpanel APIs.
  #
  # We make use of the Appstronomy::Connection class to handle the connectivity,
  # except that this class does handle the burden of loading and creating a
  # hex digest of Mixpanel API secret and token strings.
  #
  # @example
  #   service = Mixpanel::Service.new
  #   service.export(10,0, events: ["Treatment Drug Selected", "Treatment Plan Started"])
  #   service.event_names
  #   service.request('/events/names', type: 'general')
  #   service.request('/export', from_date: '2015-07-01', to_date: '2015-07-06', event: '["Treatment Plan Started", "Treatment Drug Selected"]')
  #   service.request('/events/properties/values', { event: 'Survey Completed', name: 'Practitioner Type', type: 'general', unit: 'day'})
  #
  #
  # @author Sohail Ahmed, https://github.com/idStar
  class Service

    # Maps symbols to URL path segments used to make requests on Mixpanel's API.
    API_COMMAND_PATH_MAP = {
        :event_names  => '/events/names',
        :export       => '/export'
    }

    # The number of seconds after which re-submitting the API request URI
    # should fail. We keep this low so that anyone who might intercept the URI
    # cannot actually use it to siphon off data themselves.
    #
    # You are strongly advised to keep this at `60` seconds or less. However,
    # you will likely need to increase this during any development/debugging.
    DEFAULT_REQUEST_EXPIRY_SECONDS        = 60

    # The number of seconds maximum, that scaled request expiry can grow to.
    DEFAULT_MAX_REQUEST_EXPIRY_SECONDS    = 180

    # The number of seconds to scale a request expiry, based on the date range.
    DEFAULT_SCALE_REQUEST_SECONDS_PER_DAY = 1

    # The minimum amount of time a request expiry can be set to. Applies when
    # scaling is in effect.
    MIN_REQUEST_EXPIRY_SECONDS            = 30


    BASE_API_PATH            = '/api/2.0'
    ENDPOINT_GENERAL_QUERIES = 'http://mixpanel.com'
    ENDPOINT_DATA_EXPORT     = 'https://data.mixpanel.com'



    # The convenience wrapper around using Net::HTTP that setup with
    # Mixpanel and SSL, and then keep around for subsequent invocations.
    # We'll set this up with the endpoint Mixpanel advises for data export.
    attr_reader :export_connection

    # The convenience wrapper around using Net::HTTP that setup with
    # Mixpanel and general API calls (not data export).
    # We keep this around for subsequent invocations.
    # We'll set this up with the endpoint Mixpanel advises for non-data export queries.
    attr_reader :general_connection

    # Will default to DEFAULT_REQUEST_EXPIRY_SECONDS. You can override this if you'd
    # like a shorter or longer expiry.
    attr_accessor :request_expiry_seconds


    # ---------- Initializing ----------

    # Initializes a new instance with a reusable Appstronomy::Connection
    # for each of the two Mixpanel endpoints.
    def initialize
      configure_logging
      load_configuration
      @general_connection     = Util::Connection.new(ENDPOINT_GENERAL_QUERIES)
      @export_connection      = Util::Connection.new(ENDPOINT_DATA_EXPORT)
      load_credentials
    end


    # ---------- API Operations ----------

    # Invokes an arbitrary Mixpanel API request based on the provided parameters.
    #
    # @param  [String] resource The component of the API path that is unique
    #         to this command; effectively, the web resource.
    #         e.g. '/events/names' or '/export'.
    #
    # @param  [Hash] options The query parameters we'll encode in the GET request.
    #         Generally, these let you limit or qualify the results in some manner.
    #         See the Mixpanel API reference for these options, available at:
    #         https://mixpanel.com/docs/api-documentation/data-export-api.
    #
    # @return [String] Response from Mixpanel, as a String.
    def request(resource, options = {})
      # Determine if we use the HTTPS export connection, or the general HTTP connection:
      if resource == API_COMMAND_PATH_MAP[:export] || resource == :export
        connection = @export_connection
        return_json = false
      else
        connection = @general_connection
        return_json = true
      end

      # Set a default expiry, unless the caller provided one:
      unless options[:expire]
        options[:expire] = request_expiry_timestamp
      end

      params_with_api_key = options.merge(api_key: @api_key)
      signature = digest_signature(params_with_api_key)
      params_with_signing = params_with_api_key.merge(sig: signature)

      @logger.info "Requesting resource '#{resource}' with options: #{options}"
      response = connection.get request_path(resource), params_with_signing, return_json

      # Whether we get a raw HTTP response back, or JSON inside an OpenStruct with a body key,
      # we'll want to retrieve the 'body' component of the returned response to get at the 'meat'
      # of the content.
      @logger.debug "Response: #{response.body}"
      response.body
    end


    # ---------- Conveniences ----------

    # Performs the Mixpanel data export operation. You specify a from-date
    # and to-date relative to today. The Mixpanel API uses whole days and not
    # timestamps, so we'll construct dates that look like '2015-07-09', void
    # of any concept of time.
    #
    # To get one day's worth of the latest data, you'd pass in:
    #  * from_days_ago = 1
    #  * to_days_ago = 0
    #
    # @example
    #   # Use the default of all events, for yesterday:
    #   service.export
    #
    #   # Explicitly ask for just yesterday (again, all events):
    #   service.export(1,0)
    #
    #   # Include the previous week, and just the two events specified in the events array:
    #   service.export(7,0, events: ["Survey Completed", "Navigation Action"])
    #
    #
    # @param  [String] from_days_ago The number of days back we should query for.
    # @param  [String] to_days_ago   The number of days back we should query for.
    #         Passing in zero here means today. This is useful for specifying
    #         that the end of your requested date range includes the latest
    #         data possible. Mixpanel does not return data for the present day.
    # @param  [Hash] options A hash of other query options for a data export.
    #         Notably, you can include a member `:events` that is itself an array of
    #         event names. If you do, we will properly flatten that and repackage
    #         into the query key-value pair that Mixpanel calls 'event'
    #         (even when multiple are present).
    #
    def export(from_days_ago = 1, to_days_ago = 0, options = {})
      # Record how many days a span this request is for:
      @num_days_in_request = from_days_ago - to_days_ago

      from_date_token = from_days_ago.day.ago.strftime('%Y-%m-%d')
      to_date_token = to_days_ago.day.ago.strftime('%Y-%m-%d')

      date_options = {
          from_date:  from_date_token,
          to_date:    to_date_token
      }

      options.merge! date_options

      if options[:events]
        # We need to build a string representation of a JSON array:
        events_as_json_string = options[:events].as_json.to_s

        # Note: The Mixpanel key is 'event', even though it is an array with
        # potentially multiple events present:
        options.merge!(event: events_as_json_string)

        # Now remove the 'events' array from the options, since we won't
        # be sending that to Mixpanel; just the flattened 'event' value
        # that we just created above:
        options.delete :events
      end

      request(:export, options)
    end


    # Queries Mixpanel for the list of all event names that it knows of.
    # Note that Mixpanel API will limit this request to 255 event names.
    # If you have fewer than this number events in your system, then you can
    # rest assured that you will indeed, get all of them.
    #
    # This method doesn't filter on dates; the only parameter we pass to
    # this resource request is 'general', in the hopes we get the full list
    # of event names ever used, regardless of popularity.
    def event_names
      request(:event_names, type: 'general')
    end


    private

    # Sets up our logging instance attribute, with a logger for this class.
    # By default, this class specific logger inherits settings from the root logger.
    # @return [void]
    def configure_logging
      @logger = Logging.logger[self]
    end


    # Retrieves the path to use for the operation specified. We simply pull
    # this path out of a map stored in this class. However, if we don't recognize
    # the option parameter, we'll treat it as a literal resource.
    #
    # @param  [:Symbol, String] operation Either the lookup key in our
    #         API_COMMAND_PATH_MAP of registered Mixpanel resources, or the unique resource
    #         path suffix itself, which we would combine with the BASE_API_PATH to arrive
    #         at a full resource path.
    #
    # @return The full path of a resource, minus the endpoint (protocol, host) itself.
    def request_path(operation)
      if API_COMMAND_PATH_MAP[operation]
        api_request_path = BASE_API_PATH + API_COMMAND_PATH_MAP[operation]
      else
        api_request_path = BASE_API_PATH + operation
      end

      api_request_path
    end


    def request_expiry_seconds
      if @scale_request_expiry
        num_seconds = @scale_request_seconds_per_day * (@num_days_in_request || 0)
        num_seconds = [num_seconds, MIN_REQUEST_EXPIRY_SECONDS].max
        num_seconds = [num_seconds, @max_scaled_request_expiry_seconds].min
      else
        num_seconds = @unscaled_request_expiry_seconds
      end

      @logger.info "Using request expiry of #{num_seconds.round} seconds."
      num_seconds.round
    end


    def request_expiry_timestamp
      Time.now.to_i + request_expiry_seconds
    end


    # Creates a digest signature using the API key and secret this class has already
    # retrieved.
    #
    # @param  [Hash] params The parameters to be sent to Mixpanel.
    #         We'll add the 'api_key' entry.
    # @return A digest signature that can be dropped in as the value for the query string
    #         key 'sig'.
    def digest_signature(params)
      signing_params = params.merge(api_key: @api_key)

      # We use a map function below instead of to_query on the Hash of params, since per Mixpanel,
      # we are not to include ampersands between key-value pairs. As well, we are to provide the
      # resource request arguments in alphabetical (sorted) order.
      Digest::MD5.hexdigest(
          params.map { |key, value| "#{key}=#{value}" }.sort.join + @api_secret
      )
    end


    def load_credentials
      file_path = File.join(__dir__, '../../config/MixpanelCredentials.config.json')

      unless File.exists? file_path
        @logger.fatal "Could not find file 'MixpanelCredentials.config.json' at path '#{file_path}'. \
              We need they API key and API secret for your installation from this file in order \
              to send requests to Mixpanel's API Service."
        exit!
      end

      file = File.read(file_path)
      credentials = JSON.parse(file)

      @api_key = credentials['API Key']
      @api_secret = credentials['API Secret']
    end


    # ---------- Configuration ----------

    # Sets various instance attributes related to events to include/exclude, as well
    # as request expiry timing, by consulting with the {ExporterConfig} class.
    #
    # @return [void]
    def load_configuration
      config = Util::ExporterConfig.new
      @scale_request_expiry = config.scale_request_expiry
      @scale_request_seconds_per_day = config.scale_request_seconds_per_day
      @unscaled_request_expiry_seconds = config.unscaled_request_expiry_seconds || DEFAULT_REQUEST_EXPIRY_SECONDS
      @max_scaled_request_expiry_seconds = config.max_scaled_request_expiry_seconds || DEFAULT_MAX_REQUEST_EXPIRY_SECONDS
    end

  end # class

end # module


# ---------- Development Testing ----------

#service = Mixpanel::Service.new
#service.export(10,0, events: ["Treatment Drug Selected", "Treatment Plan Started"])
#service.event_names
#service.request('/events/names', type: 'general')
#service.request('/export', from_date: '2015-07-01', to_date: '2015-07-06', event: '["Treatment Plan Started", "Treatment Drug Selected"]')
#service.request('/events/properties/values', { event: 'Survey Completed', name: 'Practitioner Type', type: 'general', unit: 'day'})